Tweaks for WPDB and %i

craigfrancis opened the pull request:
Some feedback from Juliette (@jrfnl):

Trac ticket: https://core.trac.wordpress.org/ticket/52506

---

1. Using `%1s`, you're right, it's not common, so I've replaced those examples (except the "risky" and "safer approaches" examples, to show both styles).
2. Dockblock for `_escape_identifier_value()`, I copied the "To quote the identifier itself" line directly from the [MySQL manual](https://dev.mysql.com/doc/refman/8.0/en/identifiers.html), I'm happy to change the word "quote" with "escape", I'm just not sure if a direct quote is better?
3. `@access private`, you're right, it shouldn't be used on a method.
4. With the doing it wrong line, I'd like to keep the `(%s)` in there (to show what's wrong), but I also made a mistake in saying "String and Identifier" escaping, because Integers and Floats also exist - does the suggested change help?
5. I've added 'wpdb->allow_unsafe_unquoted_parameters private' to the list of `protected_members` (so it cannot be easily/accidentally set to false; at least for now).

craigfrancis committed:
JRF: %1s is not commonly used, might be clearer to use %1$s

craigfrancis committed:
JRF: `@access private` should only be used for global functions which can't be assigned visibility

craigfrancis committed:
JRF: Trying to make the message clearer

craigfrancis committed:
Keep `wpdb->allow_unsafe_unquoted_parameters` private

jrfnl commented:
> Dockblock for `_escape_identifier_value()`, I copied the "To quote the identifier itself" line directly from the [MySQL manual](https://dev.mysql.com/doc/refman/8.0/en/identifiers.html), I'm happy to change the word "quote" with "escape", I'm just not sure if a direct quote is better?

To me (but that may just be me), it confused me when I first read it in this context (not when I read it in the MySQL manual).

The reason for the confusion is that the `prepare()` function is all about wrapping the values which replace the placeholders in the correct type of quotes (single quotes, double quotes or backticks), so the term _quoting_ in this context - to me - refers to the surrounding quotes, while this comment is about _how to use the character which is used for the wrapping quotes **within** the value itself_.

In PHP terminology, this is commonly called _escaping_ and that term is also used in the actual method name `_escape_identifier_value()`.

In that respect, the `escape_identifier()` method should maybe be called `[backtick_]quote_identifier()` ?

Also see https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.single and https://www.php.net/manual/en/regexp.reference.escape.php for examples of the "escape" terminology being used in PHP.

jrfnl commented:
For transparency, these are the comments I previously shared with @craigfrancis via Slack:

> * Example code in various docblocks often use `%1s`, which is not commonly used. Might be clearer to use `%1$s` in those examples.
> * In the docblock for `_escape_identifier_value()`: _"To quote the identifier itself, then you need to double the character, e.g. `a``b`"_ => this sentence is unclear for the average reader. I think replacing "quote" with "escape" would make it clearer.
> * `@access private` should be removed. This tag should only be used for global functions which can't be assigned visibility, not for OO methods which do have visibility. (and to be honest, I'd prefer to get rid of it completely as if something should be private, the code should be written to make it so and should not be a global function)
> * Doing it wrong: _"Arguments (%s) cannot be used for both String and Identifier escaping."_ => I saw the discussion about making the message clearer. What about _"The same argument cannot be prepared to be used as both an identifier name as well as a string value"_ ?
> * That `private` property you added to `wpdb` ? Not private at all (in practice)... have a look at the magic methods in the class...

jrfnl commented:
> With the doing it wrong line, I'd like to keep the (`%s`) in there (to show what's wrong), but I also made a mistake in saying "String and Identifier" escaping, because Integers and Floats also exist - does the suggested change help?

Current, updated, phrasing as per this PR: _"The arguments (%s) cannot be prepared as both an Identifier, and as a value."_

If I read the code correctly, the `%s` will be replaced by one or more comma-separated numbers.
Those numbers will correspond to the 0-index based number of the replacement argument(s) - in contrast to the 1-index based numbers used in the placeholder arguments.

In the tests I could only find one example of this: `WHERE id = %d AND %i LIKE %2$s LIMIT 1`, which - if I read the code correctly - would result in the following _doing it wrong_ message: _The arguments (1) cannot be prepared as both an Identifier, and as a value._

If I'm honest, I'd find that message very confusing as a dev-user anyway for a variety of reasons:
* `arguments` plural when there is only a single problem situation.
* Is the `1` 1-based or 0-based ?
* Does it refer to the placeholders or the replacement arguments ?

Once you look at the actual query code and - in this case - see there is only a single `%i`, you can extrapolate what the message means to say from there, but it takes work and won't get any easier if there are multiple %i placeholders, with or without numbers:
```sql
$query = 'SELECT * FROM %1$i WHERE %1$i.%2$i = %1$s AND %3$i = %2$d'; // Nonsense query, just to show what I mean.
```

This would result in the message:
`The arguments (0, 1) cannot be prepared as both an Identifier, and as a value.`

... which would confuse me no end.

I know this would probably require some thinking over on how this could be implemented, but the message might become clearer if the formatting placeholders which are involved in the conflict were shown in the error.
Maybe something like:
`Arguments cannot be prepared as both an Identifier, and as a value. Found the following conflicts: %1$i and %1$s, %2$i and %2$d`

What do you think @craigfrancis ?

craigfrancis committed:
Show placeholders used as Identifier and Value in _doing_it_wrong()

craigfrancis committed:
Use ReflectionProperty for allow_unsafe_unquoted_parameters test

craigfrancis commented:
Thanks @jrfnl, I've updated the patch to include your suggested error message. For performance reasons, it re-parses `$split_query` to get the formatting placeholders. This approach will also hopefully prevent any security concerns, because it's part of the `if` that always returns `NULL`. I've also updated the tests to check the contents of the error messages.

And thanks for the suggestion to use `ReflectionProperty()` for the `allow_unsafe_unquoted_parameters` test, which is needed now this property cannot be changed via `__set()`.

PS: Sorry for the delay, I kinda broke my elbow (ice skating), so doing some of this one handed :-)

craigfrancis committed:
Add arg_id to $arg_strings when position not specified

craigfrancis commented:
Ref the naming of `escape_identifier()` vs `quote_identifier()`.

I'm happy with either, and I think you're right on the dictionary definition. I don't think I gave it any thought, other than trying to match the other functions/methods that use "escape" in their name (doing a quick search though the PHP files, I'm simply reminded of the slightly flawed "magic quotes")... is there a way to find out what would be preferred?

craigfrancis committed:
Rename escape_identifier() to quote_identifier(), thanks @jrfnl

craigfrancis commented:
I've created [PR 3724](https://github.com/WordPress/wordpress-develop/pull/3724) to support the annoying `"%%%s"`... and incorporated these changes with it.
